File: src/main/kotlin/furhatos/app/arthurcompanion/state/State.kt
package furhatos.app.arthurcompanion.state

import java.time.Instant

data class State(
    val agenda: List<AgendaItem> = emptyList(),
    val happiness: Happiness = Happiness(0.6, null),
    val routines: Map<String, RoutineStatus> = emptyMap(),
    val memory: MemoryState = MemoryState(),
    val conversation: ConversationState = ConversationState(),
)

data class AgendaItem(
    val id: String,
    val text: String,
    val status: AgendaStatus = AgendaStatus.pending,
    val createdAt: Instant,
    val doneAt: Instant? = null,
)

enum class AgendaStatus {
    pending,
    done,
}

data class Happiness(
    val score: Double,
    val updatedAt: Instant?,
)

data class RoutineStatus(
    val lastDoneAt: Instant? = null,
    val lastPromptedAt: Instant? = null,
)

data class MemoryState(
    val localDate: String? = null,
    val words: List<String>? = null,
    val givenAt: Instant? = null,
    val recallCount: Int = 0,
    val lastRecallAt: Instant? = null,
)

data class ConversationState(
    val lastUserSeenAt: Instant? = null,
    val lastBotSpokeAt: Instant? = null,
)

File: src/arthurcompanion/state.py
import json
from pathlib import Path
from typing import Any, Dict

class StateStore:
    def __init__(self, path: str = "data/state.json"):
        self.path = Path(path)
        self.path.parent.mkdir(parents=True, exist_ok=True)

    def load(self) -> Dict[str, Any]:
        if not self.path.exists():
            return {
                "agenda": [],
                "happiness": {"score": 0.6, "updatedAt": None},
                "routines": {},
                "memory": {
                    "localDate": None,
                    "words": None,
                    "givenAt": None,
                    "recallCount": 0,
                    "lastRecallAt": None
                },
                "conversation": {"lastUserSeenAt": None, "lastBotSpokeAt": None}
            }
        with open(self.path, "r", encoding="utf-8") as f:
            return json.load(f)

    def save(self, state: Dict[str, Any]) -> None:
        # Standard Python IO as agreed (Option A)
        with open(self.path, "w", encoding="utf-8") as f:
            json.dump(state, f, indent=2)

    def update(self, fn) -> Dict[str, Any]:
        current = self.load()
        next_state = fn(current)
        self.save(next_state)
        return next_state
